cardano-cli/src/Cardano/CLI/Shelley/Run/Query.hs

```diff
@@ -380,6 +450,150 @@ writeLedgerState mOutFile qState@(SerialisedLedgerState serLedgerState) =
      first (const ls) (decodeFull ls)
 
 
+-- | This local data structure is introduced purely to allow nicely formatted output and has no other purpose.
+-- mark, set, go are the three ledger state stake snapshots (most recent to least recent)
+-- go is the snapshot that is used for the current epoch, set will be used in the next epoch,
+-- mark for the epoch after that.  marktot etc record the total active stake for each snapshot
+-- This information can be used by community tools to calculate upcoming leader schedules
+data Stakes =  Stakes {
+      mark, set, go :: Integer,
+      marktot, settot, gotot :: Integer
+    } deriving Show
+
+-- | Pretty printing for stake information
+instance ToJSON Stakes where
+    toJSON (Stakes m s g mt st gt) =
+        object ["poolStakeMark" .= m, "poolStakeSet" .= s, "poolStakeGo" .= g,
+                "activeStakeMark" .= mt, "activeStakeSet" .= st, "activeStakeGo" .= gt]
+
+    toEncoding  (Stakes m s g mt st gt) =
+        pairs ("poolStakeMark" .= m <> "poolStakeSet" .= s <> "poolStakeGo" .= g <>
+               "activeStakeMark" .= mt <> "activeStakeSet" .= st <> "activeStakeGo" .= gt)
+
+
+writeStakeSnapshot :: forall era ledgerera.
+                      ShelleyLedgerEra era ~ ledgerera
+                   => Era.Era ledgerera
+                   => FromCBOR (LedgerState era)
+                   => Hash StakePoolKey
+                   -> SerialisedLedgerState era
+                   -> ExceptT ShelleyQueryCmdError IO ()
+
+writeStakeSnapshot poolId qState =
+  case decodeLedgerState qState of
+    Left bs -> firstExceptT ShelleyQueryCmdHelpersError $ pPrintCBOR bs
+    Right ledgerState ->
+      if maybehk == Nothing then
+        left ShelleyQueryCmdPoolIdError
+      else
+        liftIO . LBS.putStrLn $  encodePretty $ Stakes {mark = markStake,set = setStake,go = goStake,marktot = markTotal,settot = setTotal,gotot = goTotal}
+      where
+        -- Ledger State
+        (LedgerState snapshot)  = ledgerState
+
+        -- The three stake snapshots, obtained from the ledger state
+        (SnapShots markS setS goS _) = esSnapshots $ nesEs snapshot
+
+        -- Calculate the three pool and active stake values for the given pool
+        markStake =   getPoolStake hk markS
+        setStake =    getPoolStake hk setS
+        goStake =     getPoolStake hk goS
+
+        markTotal =   getAllStake markS
+        setTotal =    getAllStake setS
+        goTotal =     getAllStake goS
+
+        -- Convert the hash string into a KeyHash for use by the ledger
+        maybehk = hashFromStringAsHex $ filter (/= '"') $ show poolId
+        hk = KeyHash $ fromJust $ maybehk
+
+ where
+   decodeLedgerState
```

Comment:
Hm...this command may hang because of the size of the ledger state. Context: https://input-output-rnd.slack.com/archives/CC0CLNEH1/p1617065989042000

Reply:
No it doesn't :).  A major point of the command is to produce a specialised and more efficient version of the ledger-state query that allows SPOs to reduce their memory requirements.   It uses a fraction of the memory.  Example call on Mainnet.

```
time getsnapshot 00beef0a9be2f6d897ed24a613cf547bb20cd282a04edfc53d477114
{
    "poolStakeGo": 40278547538358,
    "activeStakeGo": 22753958467474959,
    "poolStakeMark": 40424218559492,
    "activeStakeMark": 22670949084364797,
    "poolStakeSet": 39898761956772,
    "activeStakeSet": 22488877070796904
}

real	0m20.085s
user	0m8.874s
sys	0m2.585s

```